### ISSUE TITLE
Provide hooks for testing supporting code

### ISSUE THREAD

**Issue by geeksam** [github/scientist#95]
Before anything else:  thanks for this fantastic library!  It's very thoughtfully designed, and the README is exemplary.  The existence of this issue probably says more about me than about y'all.  :) <3

I've used Scientist to test out a few code changes now.  Rather than write to logs (which are annoying to search), I'm using a few ActiveRecord models to persist experiments, results, and observations.  (Since my work doesn't operate at roflscale like GitHub does, I have the luxury of being able to write to the DB.)  This is mostly pretty great, but I've had to commit a few sins in order to test some of my features.

Issues I've had so far:
1. If a specific experiment doesn't provide its own `#clean` block, I'd like to always use a default.  However, calling `experiment.clean.nil?` has the side effect of setting the clean block to nil.  Instead, I've had to override the `#run` method so I can run this code before calling `super`:  `@_scientist_cleaner ||= ->(value) { Experiment::DefaultCleaner.call(value) }`  **Suggested change:  add an accessor method that I can query instead of setting the ivar directly.**
2. In order to test the code that compares control/candidate timing data, I have to stub `Process.clock_gettime` with a series of ordered values, then manually compute the difference between them.  This works, but feels a bit unclean.  **Suggested change:  add an internal method (call it `#record_duration` ?) that calls the timing method, yields, sets @duration, and returns, and call this method in `Scientist::Observation#initialize`.**  Including code can then stub this method to provide its own [canned, in my case] timing data.
3. Along with #2, I also have to provide branching logic in my tests, because I don't know whether the control or candidate will run first.  Not only is this more difficult to read, it also introduces nondeterministic behavior to my test suite.  😱  **Suggested change:  extract the `behaviors.keys.shuffle` in `Scientist::Experiment#run` to an internal method (call it `#wibble` ?) that returns the list of behavior names to run in order.**  As with my suggestion for 2, this facilitates stubbing.

The following RSpec snippet illustrates items 2 and 3.  (Note:  the nested lists in the arguments to `#stub_clock` are purely for readability.)

```ruby
# Scientist uses Process.clock_gettime for timing (as it should!)
def stub_clock(*ordered_values)
  allow( Process ).to receive( :clock_gettime ) \
    .with( Process::CLOCK_MONOTONIC, :float_second ) \
    .and_return( *ordered_values.flatten )
end

specify "the saved records contain timing data" do
  # The first run should take 0.5 seconds; the second 1.0
  # Again, note that order is random
  stub_clock [ 1.0, 1.5 ], [ 2.0, 3.0 ]
  Experiment.science "wibble" do |e|
    e.use { :wibble }
    e.try { :wobble }
  end

  case control.duration
  when 0.5
    expect( candidate.duration ).to eq(  1.0 )
    expect( result.time_delta  ).to eq( -0.5 )
    expect( result.speedup     ).to eq( -2.0 )
  when 1.0
    expect( candidate.duration ).to eq(  0.5 )
    expect( result.time_delta  ).to eq(  0.5 )
    expect( result.speedup     ).to eq( +2.0 )
  else
    fail "expected control duration to be one of the two stubbed values, but was #{control.duration.inspect}"
  end
end
```

I'm happy to submit a PR for these changes, but figured I should start a discussion here first, given that:
- this mostly affects people working on [probably internal] Scientist-related infrastructure, and provides no benefit to people using Scientist as described in the README;
- my suggested solutions for 2 and especially 3 could make it easier for users to do Bad Science, in that the existence of a control will always tempt someone to use it;
- these changes would exist only for testability.  (I personally don't mind that, but recognize the existence—and possibly even the validity—of other opinions. 😜)

**Comment by geeksam** [comment]
Also, I do have "extract my AR infrastructure into a public Rails engine" on my to-do list, but it's a few items down.  Meanwhile, though, here's some spiffy output to show how I'm using Scientist (mostly to refactor nasty AR queries without inadvertently breaking correctness or performance):

```----------------------------------------------------------------------------------------------------
Experiment: live-feed-user
----------------------------------------------------------------------------------------------------
Results collected from: 2019-02-14T13:03:51-10:00
Results collected to:   2019-02-14T13:12:44-10:00

156 of 157 (99.36%) correct
  1 of 157 (0.64%) mismatched
  0 of 157 (0.00%) raised errors

Time deltas:
      0th percentile:  -13.152 sec  [*                        |                         ]   -75.6x
      5th percentile:   -3.320 sec  [                 *       |                         ]   -23.3x
     10th percentile:   -1.417 sec  [                   *     |                         ]   -17.0x
     15th percentile:   -0.627 sec  [                      *  |                         ]    -9.0x
     20th percentile:   -0.561 sec  [                       * |                         ]    -5.6x
     25th percentile:   -0.283 sec  [                        *|                         ]    -4.0x
     30th percentile:   -0.173 sec  [                        *|                         ]    -3.5x
     35th percentile:   -0.118 sec  [                        *|                         ]    -3.0x
     40th percentile:   -0.095 sec  [                        *|                         ]    -2.5x
     45th percentile:   -0.080 sec  [                        *|                         ]    -2.2x
     50th percentile:   -0.066 sec  [------------------------*+-------------------------]    -1.9x
     55th percentile:   -0.049 sec  [                        *|                         ]    -1.8x
     60th percentile:   -0.039 sec  [                        *|                         ]    -1.6x
     65th percentile:   -0.035 sec  [                         *                         ]    -1.5x
     70th percentile:   -0.029 sec  [                         *                         ]    -1.4x
     75th percentile:   -0.025 sec  [                         *                         ]    -1.3x
     80th percentile:   -0.016 sec  [                         *                         ]    -1.3x
     85th percentile:   -0.012 sec  [                         *                         ]    -1.2x
     90th percentile:   -0.003 sec  [                         *                         ]    -1.1x
     95th percentile:   +1.251 sec  [                         |  *                      ]    +8.3x
    100th percentile:  +13.118 sec  [                         |            *            ]   +39.9x
----------------------------------------------------------------------------------------------------```

(Scaling is linear, not logarithmic, because I'm too lazy to math.)

**Comment by zerowidth** [comment]
> I'm happy to submit a PR for these changes, but figured I should start a discussion

As you may have guessed from my delayed response, I don't have much time available to maintain this library. However, your issue is exemplary and your suggestions reasonable! 

Were you to open a PR with those changes, I'd only ask that you include a note on testing for the README, so future testers can know what internals they should consider stubbing. Those stub-hook-worthy-methods should also be documented internally (in the tomdoc/comments) to note that they're part of a "public testing" contract so we don't break compatibility with future changes.

The ascii output for timing data looks real nice! What are you using to store the experiment results?

**Comment by geeksam** [comment]
Thanks!  I'll go ahead when I get a chance (or need a break), then.

Having forgotten about my original suggestions, I wonder if it might make sense to address items 2 and 3 by adding explicit API support for overriding timing data with something like:  `experiment.stub_timings_for_tests(control: 0.5, candidate: 0.75)`  I'll keep that in mind when I start in on the PR.  If I can't decide on one or the other, it's probably not too difficult to prepare two different PRs and let you choose.  :)

As for storing experiment results, they're just going into ActiveRecord.  I hope to extract that into a Rails engine at some point.

**Comment by geeksam** [comment]
@zerowidth I've submitted a PR with an initial sketch of what this might look like.  This is a minimally invasive diff; I didn't bother adding a `#record_duration` method as I had originally suggested.

I tried to name the new hook method in a way that should discourage its casual use while still being thematic (`#fabricate_durations_for_testing_purposes`).  I don't want to go full `#i_suck_and_my_tests_are_order_dependent`, but if you think this name should be more alarming and/or annoying to type, I have very little attachment to it.  (I'm mostly asking because I realized you might find it easier to look at the PR if I asked for specific feedback, rather than just "WDYT?", and hey, if you've read this far, mission accomplished!  😜)

**Comment by zerowidth** [comment]
I think `fabricate_durations_for_testing_purposes` is plenty annoying to type, and still clear what it's for! 😁No need for shaming developers a la `i_suck_and_my_tests_are_order_dependent`.

**Comment by geeksam** [comment]
Addressed in #103. Thanks; this was fun! :)

**Comment by geeksam** [comment]
FYI, I finally extracted the tools I used to build the purdy chart above into https://github.com/RealGeeks/lab_tech
